### PR TITLE
Implementation of data-reconciliation-app ingestion endpoint

### PR DESCRIPTION
**Pull request by Aymalla**
- Scaffolding the Data Reconciliation app #83
-  Implement and unit test data ingest APIs #82 
-  Functional test data ingest API #84

### CODE REVIEW

File: data-reconciliation-app/src/app.json [hunk 1]
```diff
@@ -1,21 +1,53 @@
...
+    "/ingest": {
       "post": {
-        "js_module": "endpoints/sample.js",
-        "js_function": "write",
-        "forwarding_required": "sometimes",
-        "authn_policies": [],
+        "js_module": "endpoints/ingest.js",
+        "js_function": "submitData",
+        "forwarding_required": "always",
+        "authn_policies": ["member_cert"],
```
Comment (ross-p-smith) [inline data-reconciliation-app/src/app.json:8]: Should this be user_cert?

File: data-reconciliation-app/test/unit-test/setup.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import "@microsoft/ccf-app/polyfill";
+import * as ccfapp from "@microsoft/ccf-app";
+
+beforeEach(function () {
+  // clear KV before each test
```
Comment (ross-p-smith) [inline data-reconciliation-app/test/unit-test/setup.ts:5]: Is it worth adding a // TODO:

File: data-reconciliation-app/src/endpoints/reporting.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  request: ccfapp.Request<any>
+): ccfapp.Response<any> {
+  try {
+    const getCallerId = authenticationService.getCallerId(request);
+    if (getCallerId.failure) {
+      return ApiResult.Response(getCallerId);
+    }
+
+    const callerId = getCallerId.content;
+    const isMember = authenticationService.isMember(callerId);
```
Comment (ross-p-smith) [inline data-reconciliation-app/src/endpoints/reporting.ts:16]: Should we use isUser?

**Comment by ross-p-smith** [comment]
Looks great to me. When we run this in docker we have to setup an extra member in setup_governance. This is different to the Sandbox as you can just say InitialMemberCount=3

In file `/workspaces/ccf-app-samples/data-reconciliation-app/governance/scripts/setup_governance.sh` on line 108 if you add 

```bash
cert_name="member2"
create_certificate "${cert_name}" "${certs}"
$root_dir/scripts/add_member.sh --cert-file $certs/${cert_name}_cert.pem --pubk-file $certs/${cert_name}_enc_pubk.pem
```

It will all work

**Review by ross-p-smith — CHANGES_REQUESTED** [review]
If we get this member added I am more than happy to approve!!

File: data-reconciliation-app/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
+/*
```
Comment (macromania) [inline data-reconciliation-app/jest.config.js:1]: I suggest removing all these comments and simplify the file like below:

```suggestion
module.exports = {
    transform: {
        '^.+\\.ts?$': 'ts-jest',
    },
    testEnvironment: 'node',
    testRegex: './src/.*\\.(test|spec)?\\.(ts|ts)$',
    moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
    roots: ['<rootDir>/src'],
};
```

File: data-reconciliation-app/package.json [hunk 5]
```diff
@@ -19,15 +20,19 @@
     "protobufjs": "^6.10.1"
   },
   "devDependencies": {
+    "@jest/globals": "^29.3.1",
```
Comment (macromania) [inline data-reconciliation-app/package.json:23]: Let's add following dependencies for better developer productivity:

```json
"@types/jasmine": "^4.3.0",
"@types/mocha": "^10.0.0",
"@types/node": "^18.11.9",
"ts-node": "^10.9.1",
```

Also, I'm not sure about all other dev dependencies 🤔 
Do we use these anywhere in our sample app?

```json
@jest/globals
@rollup/plugin-commonjs
@rollup/plugin-node-resolve
@rollup/plugin-typescript
@types/jsrsasign
@types/lodash-es
del-cli
http-server
rollup
```

File: data-reconciliation-app/src/app.json [hunk 6]
```diff
@@ -1,21 +1,53 @@
...
-        "openapi": {}
-      },
+    "/ingest": {
       "post": {
-        "js_module": "endpoints/sample.js",
-        "js_function": "write",
-        "forwarding_required": "sometimes",
-        "authn_policies": [],
+        "js_module": "endpoints/ingest.js",
+        "js_function": "submitData",
```
Comment (macromania) [inline data-reconciliation-app/src/app.json:6]: Everything this endpoint receives data. Removing it from the function name could look cleaner.
Given the endpoint is REST-ish and it is called `/ingest`, the method can be `postHandler`?

This could read as: `HTTP POST handler of ingest endpoint`?

File: data-reconciliation-app/src/app.json [hunk 7]
```diff
@@ -1,21 +1,53 @@
...
+        "forwarding_required": "always",
+        "authn_policies": ["member_cert"],
         "mode": "readwrite",
```
Comment (macromania) [inline data-reconciliation-app/src/app.json:9]: This endpoint only writes?

File: data-reconciliation-app/src/app.json [hunk 8]
```diff
@@ -1,21 +1,53 @@
...
+              }
+            }
+          }
+        }
+      }
+    },
+    "/report": {
+      "get": {
+        "js_module": "endpoints/reporting.js",
+        "js_function": "getDataSummary",
```
Comment (macromania) [inline data-reconciliation-app/src/app.json:35]: Similar to above suggestion, can this method called `getHandler`?

File: data-reconciliation-app/src/endpoints/ingest.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import * as ccfapp from "@microsoft/ccf-app";
+import { ServiceResult } from "../utils/service-result";
+import { ingestService, authenticationService } from "../utils/dependencies";
+import { ApiResult } from "../utils/api-result";
+import { DataSchema } from "../models/data-schema";
+
+export function submitData(request: ccfapp.Request<any>): ccfapp.Response<any> {
+  try {
```
Comment (macromania) [inline data-reconciliation-app/src/endpoints/ingest.ts:8]: This `try/catch` block is too big. The `catch` can contain multiple failures and it will be hard to understand which failure is triggered on which condition. 

I'd suggest to organise this block more like below. Also, I removed some of `{ }` for single line expressions to make code easier to read. This can be debatable on code style 😅 

```typescript

// CHANGE 1: `authenticationService` is returning a `ServiceResult`. This means, it should not throw any exception and rather handle all of them by returning a `failure`

const getCallerId = authenticationService.getCallerId(request);
if (getCallerId.failure) return ApiResult.Response(getCallerId);
const callerId = getCallerId.content!;

// CHANGE2: `authenticationService` is returning a `ServiceResult`. This means, it should not throw any exception and rather handle all of them by returning a `failure`

const isMember = authenticationService.isMember(callerId);
if (isMember.failure || !isMember.content) return ApiResult.Response(ServiceResult.Unauthorized());


// CHANGE 3: request.body.json() could throw exception. Let's catch that only
let data: object | null = null;

try {
    const data = request.body.json();
}
catch (e: unknown) {
   const response = ServiceResult.Failed({
      errorMessage: ex.message,
      errorType: "InvalidJSON",
      details: ex,
    });
    return ApiResult.Response(response);
}

// CHANGE 4: `DataSchema.mapDataRecords` should return ServiceResult.

const mapDataRecords = DataSchema.mapDataRecords(data);
if (mapDataRecords.failure) return ApiResult.Response(mapDataRecords);

const dataRecords = mapDataRecords.content!
const response = ingestService.submitData(callerId, dataRecords);
return ApiResult.Response(response);
````

File: data-reconciliation-app/src/endpoints/reporting.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import * as ccfapp from "@microsoft/ccf-app";
+import { ServiceResult } from "../utils/service-result";
+import { reportingService, authenticationService } from "../utils/dependencies";
+import { ApiResult } from "../utils/api-result";
+
+export function getDataSummary(
+  request: ccfapp.Request<any>
```
Comment (macromania) [inline data-reconciliation-app/src/endpoints/reporting.ts:7]: We can update `prettier line width` to 240. We don't have to write the method declaration in this format 🐱 
```suggestion
  export function getDataSummary(request: ccfapp.Request<any>): ccfapp.Response<any>
```

File: data-reconciliation-app/src/endpoints/ingest.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import * as ccfapp from "@microsoft/ccf-app";
+import { ServiceResult } from "../utils/service-result";
+import { ingestService, authenticationService } from "../utils/dependencies";
+import { ApiResult } from "../utils/api-result";
+import { DataSchema } from "../models/data-schema";
+
+export function submitData(request: ccfapp.Request<any>): ccfapp.Response<any> {
```
Comment (macromania) [inline data-reconciliation-app/src/endpoints/ingest.ts:7]: The response type can be `ApiResult<T>` where T can be an interface returned from `ingestService.submitData`.

```typescript
export interface SubmitDataResponse {
 //////
}

File: data-reconciliation-app/src/endpoints/reporting.ts (hunk 10 above)
Comment (macromania) [reply data-reconciliation-app/src/endpoints/reporting.ts:7]: The response type can be `ccfapp.Response<T>` where T can be an interface returned from `reportingService.getData`.

```typescript
export interface ReportingResponse {
 //////
}

File: data-reconciliation-app/src/endpoints/reporting.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+): ccfapp.Response<any> {
+  try {
```
Comment (macromania) [inline data-reconciliation-app/src/endpoints/reporting.ts:9]: Please see my comment above. The try/catch is best for dependencies and expressions that can throw an exception. Any expression returning a ServiceResult doesn't require to be inside the try/catch block. ServiceResult pattern is built on the convention of service handling all the known exceptions in the control flow.

File: data-reconciliation-app/src/models/recon-record.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+export class ReconciledRecord {
```
Comment (macromania) [inline data-reconciliation-app/src/models/recon-record.ts:1]: File name is better set as `reconciled-record.ts`?

File: data-reconciliation-app/src/models/recon-record.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
...
+  key: string;
+  values: Object = {};
```
Comment (macromania) [inline data-reconciliation-app/src/models/recon-record.ts:3]: Should we think about a proper type instead of `Object`?

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import * as ccfapp from "@microsoft/ccf-app";
+
+export interface IKeyValueRepository<T> {
+  set(key: string, value: T): T;
+  get(key: string): T;
+  has(key: string): boolean;
+  keys(): string[];
+  values(): T[];
+  get size(): number;
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:9]: Why did you need to use `get` here in the interface definition?

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// generic key-value repository wrapping ccf TypedKvMap interaction
+export class KeyValueRepository<T> implements IKeyValueRepository<T> {
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:13]: Instead of exporting the class, we better export the singleton instance of it for better testability.

I'd organise this more like:

```typescript
export interface Repository<T> {
 ///////
}

class KeyValueRepository<T> implements Repository<T> {
     private readonly private kvStore: ccfapp.TypedKvMap<string, T>;

     constructor(kvStore: ccfapp.TypedKvMap<string, T>) {
         this.kvStore = kvStore;
     }
     ////////
}

const kvStore = ccfapp.typedKv("data", ccfapp.string, ccfapp.json<T>());
const keyValueRepository: Repository = new KeyValueRepository(kvStore);

export  keyValueRepository;
```

This above suggestion aims:

1. Exporting the interface as contract instead of concrete implementation (SOLID principales)
2. Helps spying on the exported singleton instance for unit testing
3. Injects the infrastructure dependency, `kvStore` at the constructor.

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+  private kvStore: ccfapp.TypedKvMap<string, T>;
+
+  public constructor() {
+    this.kvStore = this.getDataMap();
+    console.log(this.kvStore);
+  }
+
+  // update key-value pair or create new record if key not exists
+  public set(key: string, value: T): T {
+    this.kvStore.set(key, value);
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:23]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    return value;
+  }
+
+  // retrieve key value from kv-store
+  public get(key: string): T {
+    return this.kvStore.get(key);
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:29]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  // check if key exists
+  public has(key: string): boolean {
+    return this.kvStore.has(key);
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:34]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  // retrieve all keys of kv-store
+  public keys(): string[] {
+    const keys: string[] = [];
+    this.kvStore.forEach((val, key) => {
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:40]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+      keys.push(key);
+    });
+    return keys;
+  }
+
+  // retrieve all values of kv-store
+  public values(): T[] {
+    const values: T[] = [];
+    this.kvStore.forEach((val, key) => {
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:49]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+      values.push(val);
+    });
+    return values;
+  }
+
+  // clear all key-value pairs of kv-store
+  public clear(): void {
+    this.kvStore.clear();
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:57]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/repositories/kv-repository.ts [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  // get key-value store item count
+  public get size(): number {
+    return this.kvStore.size;
```
Comment (macromania) [inline data-reconciliation-app/src/repositories/kv-repository.ts:62]: The repository should handle any exceptions thrown from `kvStore` and return `ServiceResult`

File: data-reconciliation-app/src/services/authentication-service.ts [hunk 24]
```diff
@@ -0,0 +1,68 @@
+import * as ccfapp from "@microsoft/ccf-app";
+import { ccf } from "@microsoft/ccf-app/global";
+import { User } from "../models/user";
+import { ServiceResult } from "../utils/service-result";
+
+interface Caller {
+  id: string;
+}
+
+export interface IAuthenticationService {
```
Comment (macromania) [inline data-reconciliation-app/src/services/authentication-service.ts:10]: I'd organise this like above suggestion:

```typescript
export interface AuthenticationService {
//////
}

class CertificationBasedAuthenticationService implements AuthenticationService {
     ////
}

const authenticationService: AuthenticationService = new CertificationBasedAuthenticationService();
export default authenticationService
```

File: data-reconciliation-app/src/services/authentication-service.ts [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// Authentication Service
+export class AuthenticationService implements IAuthenticationService {
+  // get caller id
+  // Note that the following way of getting caller ID doesn't work for 'jwt' auth policy and 'no_auth' auth policy.
+  public getCallerId(request: ccfapp.Request<any>): ServiceResult<User> {
+    try {
+      const caller = request.caller as unknown as Caller;
+      return ServiceResult.Succeeded(caller.id);
```
Comment (macromania) [inline data-reconciliation-app/src/services/authentication-service.ts:23]: Caller Id is string. Can we check if `caller.id` is null or undefined before this?

File: data-reconciliation-app/src/services/authentication-service.ts [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+
+  // Check if user exists https://microsoft.github.io/CCF/main/audit/builtin_maps.html#users-info
+  public isUser(userId: string): ServiceResult<boolean> {
+    try {
+      const usersCerts = ccfapp.typedKv(
+        "public:ccf.gov.users.certs",
+        ccfapp.arrayBuffer,
+        ccfapp.arrayBuffer
+      );
+      return ServiceResult.Succeeded(usersCerts.has(ccf.strToBuf(userId)));
```
Comment (macromania) [inline data-reconciliation-app/src/services/authentication-service.ts:41]: Can we first see if `usersCerts.has(ccf.strToBuf(userId))` returns true before returning `ServiceResult.Succeeded`?

File: data-reconciliation-app/src/services/authentication-service.ts [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+
+  // Check if member exists https://microsoft.github.io/CCF/main/audit/builtin_maps.html#users-info
+  public isMember(memberId: string): ServiceResult<boolean> {
+    try {
+      const membersCerts = ccfapp.typedKv(
+        "public:ccf.gov.members.certs",
+        ccfapp.arrayBuffer,
+        ccfapp.arrayBuffer
+      );
+      return ServiceResult.Succeeded(membersCerts.has(ccf.strToBuf(memberId)));
```
Comment (macromania) [inline data-reconciliation-app/src/services/authentication-service.ts:59]: Can we first see if `membersCerts.has(ccf.strToBuf(memberId))` returns true before returning ServiceResult.Succeeded?

File: data-reconciliation-app/src/services/ingest-service.ts [hunk 28]
```diff
@@ -0,0 +1,38 @@
+import { DataRecord, DataRecordMap } from "../models/data-record";
+import { User } from "../models/user";
+import { ServiceResult } from "../utils/service-result";
+import { IKeyValueRepository } from "../repositories/kv-repository";
+import { DataSchema, DataFieldSchema } from "../models/data-schema";
+import * as ccfapp from "@microsoft/ccf-app";
+
+export interface IIngestService {
```
Comment (macromania) [inline data-reconciliation-app/src/services/ingest-service.ts:8]: I'd organise this again like below 😅 

```typescript
export interface IngestService {
 //////
}

class IngestServiceImpl implements IngestService {
     private keyValueRepo: IKeyValueRepository<DataRecordMap>;
     constructor(keyValueRepo: IKeyValueRepository<DataRecordMap>) {
           this.keyValueRepo = keyValueRepo;
    }

    ////
}


const ingestService: IngestService = new IngestServiceImpl(repository);
export default ingestService;
```

File: data-reconciliation-app/src/services/reporting-service.ts [hunk 29]
```diff
@@ -0,0 +1,23 @@
+import { DataRecord, DataRecordMap } from "../models/data-record";
+import { User } from "../models/user";
+import { ServiceResult } from "../utils/service-result";
+import { IKeyValueRepository } from "../repositories/kv-repository";
+import { ReconciledRecord } from "../models/recon-record";
+
+export interface IReportingService {
```
Comment (macromania) [inline data-reconciliation-app/src/services/reporting-service.ts:7]: Please see above how to organise this class 🤣

File: data-reconciliation-app/src/utils/api-result.ts [hunk 30]
```diff
@@ -0,0 +1,16 @@
+import { ServiceResult } from "./service-result";
+
+export interface CCFResult {
+  statusCode: number;
+  body: any;
+}
+
+export class ApiResult {
+  public static Response<T>(result: ServiceResult<T>): CCFResult {
```
Comment (macromania) [inline data-reconciliation-app/src/utils/api-result.ts:9]: Can we organise this to `ApiResult.Success`, `ApiResult.Failure`, `ApiResult.AuthFailure` to provide better context for the code reader?

File: data-reconciliation-app/src/utils/dependencies.ts [hunk 31]
```diff
@@ -0,0 +1,30 @@
+import {
```
Comment (macromania) [inline data-reconciliation-app/src/utils/dependencies.ts:1]: I'm not sure about pulling all dependencies in single file. Every module should decide their dependencies. Otherwise, this file will be a burden to maintain.

File: data-reconciliation-app/src/utils/service-result.ts [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+
+  public static Succeeded<T>(content: T): ServiceResult<T> {
+    return new ServiceResult<T>(content, null, true, StatusCode.OK);
+  }
+
+  public static Failed<T>(error: ErrorResponse): ServiceResult<T> {
+    return new ServiceResult<T>(null, error, false, StatusCode.BAD_REQUEST);
+  }
+
+  public static Unauthorized<T>(
```
Comment (macromania) [inline data-reconciliation-app/src/utils/service-result.ts:43]: `Unauthorized` is a concern at the endpoint level by looking at the request. This Result pattern class is more universal. I'm not sure having this method here is fit for purpose while we have `Failed` covering almost all failures.

File: data-reconciliation-app/src/utils/service-result.ts [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+    this.statusCode = statusCode;
+    this.status = success ? "Success" : "Error";
+  }
+
+  public static Succeeded<T>(content: T): ServiceResult<T> {
+    return new ServiceResult<T>(content, null, true, StatusCode.OK);
+  }
+
+  public static Failed<T>(error: ErrorResponse): ServiceResult<T> {
+    return new ServiceResult<T>(null, error, false, StatusCode.BAD_REQUEST);
```
Comment (macromania) [inline data-reconciliation-app/src/utils/service-result.ts:40]: I think `StatusCode` here is too much of a concern for a `ServiceResult`. Maybe it is better at `APIResult`?

File: data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts [hunk 34]
```diff
@@ -0,0 +1,60 @@
...
+
+import * as jscrypto from "crypto";
+import {
+  IKeyValueRepository,
+  KeyValueRepository,
+} from "../../../src/repositories/kv-repository";
+import { DataRecord } from "../../../src/models/data-record";
+import { describe, expect, test, beforeEach, afterEach } from "@jest/globals";
+
+describe("Key value pair Repository", () => {
```
Comment (macromania) [inline data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts:11]: Our repository implementation is a wrapper/abstraction on `kyStore`. By using `jest.Spy`, we could avoid the `polyfill`?

File: data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts [hunk 35]
```diff
@@ -0,0 +1,60 @@
...
+    keyValueRepo = new KeyValueRepository<DataRecord>();
+  });
+
+  afterEach(() => {});
+
+  test("Should add a new key-value pair", () => {
+    // Act
+
+    // Assert
+    const result = keyValueRepo.set(testKey, testDataRecord);
```
Comment (macromania) [inline data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts:27]: I didn't see any business logic at `Repository` which just uses the `kvStore` as is. Do we really need to test it?

File: data-reconciliation-app/test/unit-test/services/ingest-service.test.ts [hunk 36]
```diff
@@ -0,0 +1,42 @@
+import { describe, expect, test, beforeEach, afterEach } from "@jest/globals";
+import * as jscrypto from "crypto";
```
Comment (macromania) [inline data-reconciliation-app/test/unit-test/services/ingest-service.test.ts:2]: You don't have to use this at all. NodeJS has built in method for this :) 

```
import { randomUUID } from 'crypto';

const userId = randomUUID();
```

Remember to add following types to your `devDependencies`

```json
"@types/node": "^18.11.9",
```

File: data-reconciliation-app/tsconfig.json [hunk 37]
```diff
@@ -3,12 +3,14 @@
     "lib": ["ES2020"],
     "target": "ES2020",
     "module": "ES2020",
+    "esModuleInterop": true,
```
Comment (macromania) [inline data-reconciliation-app/tsconfig.json:6]: I suggest following `tsconfig` for maximum backwards compatibility and developer productivity:

```json
{
    "compilerOptions": {
        "target": "ES2018",
        "module": "commonjs",
        "lib": [
            "es2018"
        ],
        "declaration": true,
        "resolveJsonModule": true,
        "strict": true,
        "noImplicitAny": false,
        "strictNullChecks": true,
        "noImplicitThis": true,
        "alwaysStrict": true,
        "noUnusedLocals": false,
        "noUnusedParameters": false,
        "noImplicitReturns": true,
        "noFallthroughCasesInSwitch": false,
        "inlineSourceMap": false,
        "inlineSources": false,
        "experimentalDecorators": true,
        "strictPropertyInitialization": false,
        "esModuleInterop": true,
        "skipLibCheck": true,
        "sourceMap": true,
        "outDir": "build",
        "typeRoots": [
            "./node_modules/@types"
        ],
        "types": [
            "node",
            "jasmine",
            "mocha",
            "jest"
        ]
    },
    "include": [
        "**/*"
    ],
    "exclude": [
        "./build/**/*",
        "docs"
    ]
}
```

**Review by macromania — CHANGES_REQUESTED** [review]
Most of my suggestions are towards the code style and conventions from our `typescript-starter` project. Some of it also to improve OO design and testability. Some more added for developer productivity and IDE integration.

File: data-reconciliation-app/src/models/data-record.ts [hunk 38]
```diff
@@ -0,0 +1,18 @@
+export type DataAttributeType = string | number;
+export class DataRecord {
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply data-reconciliation-app/src/models/data-record.ts:2]: Lets use IDataRecord though and not mix and match naming convention (we are using I instead of Props elsewhere)

[the remark this answers is not in the document]
Comment (macromania) [reply data-reconciliation-app/src/models/data-record.ts:2]: Actually, `Props` here is used to define what `Props` this class is needed at the constructor. So, it is a definition for DTO and not Service.

File: data-reconciliation-app/src/app.json (hunk 7 above)
Comment (Aymalla) [reply data-reconciliation-app/src/app.json:9]: It will perform read and update operations to key-value store

File: data-reconciliation-app/src/repositories/kv-repository.ts (hunk 15 above)
Comment (Aymalla) [reply data-reconciliation-app/src/repositories/kv-repository.ts:9]: To export it as property call style "repo.size"

File: data-reconciliation-app/src/services/authentication-service.ts (hunk 27 above)
Comment (Aymalla) [reply data-reconciliation-app/src/services/authentication-service.ts:59]: I think it doesn't matter here because true or false are valid results, and 'false' doesn't mean failure

File: data-reconciliation-app/src/utils/dependencies.ts (hunk 31 above)
Comment (Aymalla) [reply data-reconciliation-app/src/utils/dependencies.ts:1]: This file act as Startup.cs , where all dependencies are injected

File: data-reconciliation-app/src/repositories/kv-repository.ts (hunk 17 above)
Comment (Aymalla) [reply data-reconciliation-app/src/repositories/kv-repository.ts:23]: I think we don't expect many exceptions from kvstore, and if it happens, we will handle it on the API layer.
Beside that, returning the value directly from the repository will give simple and readable code, which will be suitable for the samples repository.

How does it sound?

File: data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts (hunk 34 above)
Comment (Aymalla) [reply data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts:11]: There is an issue when you call any function related to ccf-kvstore and it throws an exception. so the ccf-kvstore need to be mocked

```
Test suite failed to run > Jest encountered an unexpected token
/workspaces/aym-ccf-app-samples/data-reconciliation-app/node_modules/@microsoft/ccf-app/index.js:16
    export * from "./kv.js";

    SyntaxError: Unexpected token 'export'
```

File: data-reconciliation-app/package.json (hunk 5 above)
Comment (Aymalla) [reply data-reconciliation-app/package.json:23]: Yeah it is used for build the app bundle that will be deployed to CCF

File: data-reconciliation-app/src/app.json (hunk 1 above)
Comment (Aymalla) [reply data-reconciliation-app/src/app.json:8]: I've modified it to accept calls from both members and users; later, I will change the test cases to test both scenarios.

File: data-reconciliation-app/src/endpoints/reporting.ts (hunk 3 above)
Comment (Aymalla) [reply data-reconciliation-app/src/endpoints/reporting.ts:16]: I've modified it to accept calls from both members and users; later, I will change the test cases to test both scenarios.

File: data-reconciliation-app/tsconfig.json (hunk 37 above)
Comment (Aymalla) [reply data-reconciliation-app/tsconfig.json:6]: It causes some issues, I will do it as seperate PR

File: data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts (hunk 35 above)
Comment (Aymalla) [reply data-reconciliation-app/test/unit-test/repositories/kv-repository.test.ts:27]: It is testing the integration with ccf key-value map

**Comment by Aymalla** [comment]
> If we get this member added I am more than happy to approve!!

Done

File: data-reconciliation-app/src/models/data-record.ts (hunk 38 above)
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply data-reconciliation-app/src/models/data-record.ts]: Feel free to ignore my comment then!

File: data-reconciliation-app/src/repositories/kv-repository.ts (hunk 17 above)
Comment (macromania) [reply data-reconciliation-app/src/repositories/kv-repository.ts:23]: Handling exception at the place of happening is much more readable code and much more maintainable. When an exception bubbles up all the way to API layer, that is an invitation for bad SLA of the service.

I highly recommend handling it here for that reason :)

File: data-reconciliation-app/src/utils/dependencies.ts (hunk 31 above)
Comment (macromania) [reply data-reconciliation-app/src/utils/dependencies.ts]: I think there is no `Startup.cs` mentality in NodeJS. The module system and correspondingly the tree-shaking for bundling, reducing size etc are all built on simple conventions. The NodeJS module system doesn't require a Dependency Injection framework at all 😃

File: data-reconciliation-app/src/repositories/kv-repository.ts (hunk 17 above)
Comment (Aymalla) [reply data-reconciliation-app/src/repositories/kv-repository.ts]: I feel it makes the code more focused on checking the result of each call (to inspect the exceptions) instead of trusting the provided implementation. From my point of view, I think it is better to treat exceptions as exceptional and handle them on a certain layer, excluding business logic validation error from that.

I will implement it as you mention, and later we can have a deeper discussion.

File: data-reconciliation-app/src/repositories/kv-repository.ts (hunk 23 above)
Comment (Aymalla) [reply data-reconciliation-app/src/repositories/kv-repository.ts]: done

File: data-reconciliation-app/src/repositories/kv-repository.ts (hunk 16 above)
Comment (Aymalla) [reply data-reconciliation-app/src/repositories/kv-repository.ts]: Yeah, I agree; I already consume the interface. But exporting the concrete class to make it available for unit testing

File: data-reconciliation-app/src/utils/dependencies.ts (hunk 31 above)
Comment (Aymalla) [reply data-reconciliation-app/src/utils/dependencies.ts]: Yeah, but more rich typescript frameworks like angular using Dependency Injection.

I will reomve it as you mention, and later we can have a deeper discussion.

**Review by macromania — APPROVED** [review]
Great work 👏🏼  and thank you for taking my feedback into consideration for your overall implementation.
